Sources/App/Views/PackageController/PackageShowView.swift

```diff
@@ -288,13 +193,186 @@ extension PackageShowView.Model {
+                lpInfo[keyPath: $0].platforms == v.platforms
+                    && lpInfo[keyPath: $0].swiftVersions == v.swiftVersions }
+            availableKeyPaths.removeAll(where: { group.contains($0) })
+            return group
+        }
+        .filter { !$0.isEmpty }
+        return groups
+    }
+
+    func lpInfoSection(for keypaths: [LanguagePlatformKeyPath]) -> [Node<HTML.BodyContext>] {
+        guard let leadingKeyPath = keypaths.first else { return [] }
+        let cssClasses: [LanguagePlatformKeyPath: String] = [\.stable: "stable",
+                                                             \.beta: "beta",
+                                                             \.latest: "branch"]
+        let nodes = keypaths.map { kp -> Node<HTML.BodyContext> in
+            let info = languagePlatforms[keyPath: kp]
+            let cssClass = cssClasses[kp]!
+            return .a(
+                .href(info.link.url),
+                .span(
+                    .class(cssClass),
+                    .i(.class("icon \(cssClass)")),
+                    .text(info.link.name)
+                )
+            )
+        }
+
+        // swift versions and platforms are the same all versions because we grouped them,
+        // so we use the leading keypath to obtain it
+        let versionInfo = languagePlatforms[keyPath: leadingKeyPath]
+
+        return [
+            .p(
+                .group(Self.listPhrase(opening: .text("Version".pluralized(for: keypaths.count) + " "),
+                                       nodes: nodes)),
+                " \("supports".pluralized(for: keypaths.count, plural: "support")):"
+            ),
+            .ul(
+                .li(
+                    .group(versionsClause(versionInfo.swiftVersions))
+                ),
+                .li(
+                    .group(platformsClause(versionInfo.platforms))
+                )
+            )
+        ]
+    }
+
+    func versionsClause(_ versions: [String]) -> [Node<HTML.BodyContext>] {
+        let nodes = versions.map { Node<HTML.BodyContext>.strong(.text($0)) }
+        return Self.listPhrase(opening: .text("Swift "), nodes: nodes)
+    }
+
+    func platformsClause(_ platforms: [Platform]) -> [Node<HTML.BodyContext>] {
+        let nodes = platforms
+            .sorted(by: { $0.ordinal < $1.ordinal })
+            .map { "\($0)+" }
+            .map { Node<HTML.BodyContext>.strong(.text($0)) }
+        return Self.listPhrase(opening: .text(""), nodes: nodes)
+    }
+
+}
+
+
+// MARK: - General helpers
+
+extension PackageShowView.Model {
+
+    static func listPhrase(opening: Node<HTML.BodyContext>,
+                           nodes: [Node<HTML.BodyContext>],
+                           ifNoValues: [Node<HTML.BodyContext>] = []) -> [Node<HTML.BodyContext>] {
+        switch nodes.count {
+            case 0:
+                return ifNoValues
+            case 1:
+                return [opening, nodes[0]]
```

Comment:
Subtle one here, and one that complicates the solution a bit. There shouldn't be a comma for just two items.

* One item.
* One item and another item.
* One item, another item, and a final item.

Reply:
That's how it shoud work. Only the default case adds any commas (in the `middle` section).

I don't have an explicit test for this because the nodes are hard to compare but the snapshot should confirm it - or is this not covering those cases?

Reply:
I'm sure you're right, I was just eyeballing the code and I didn't see a difference with the commas in the code. Apologies!